README
Parse-Readelf version 0.17
==========================

Parse::Readelf is a module to parse the output of readelf from the GNU
binutils suite.  At the moment only small parts of the debugging
section are supported to get information about the structure layout of
the variables of a binary object.  Also only Dwarf-2 debugging
information is understood.

INSTALLATION

To build, test and install this module, run the following commands:

    perl Build.PL
    perl Build
    perl Build test
    perl Build install

If you prefer the traditional way of installation, use the following:

    perl Makefile.PL
    make
    make test
    make install

DEPENDENCIES

As this module parses the output of readelf of course you need that.
objdump would work as well but as it's part of the same suite there is
no real reason to use it instead of readelf (at least for Dwarf-2, for
Stabs or other debugging formats there might be a difference).

The test-suite needs the Unix command "cat" to simulate a test run
with prepared output.

In addition you'll need gcc (g++) if you like to test the examples in
the "Examples" directory.

Other than that the example script "run_examples.pl" needs a shell
supporting I/O redirection.  If you don't have one, you probabely have
to try the examples by hand.

This module requires these other standard modules and libraries:

    Carp
    File::Spec
    Test::More

BUGS

This is quite incomplete yet.  I only worked on that what I needed
most which was the memory layout of variables and data types.

Also note that though the self-tests almost completely cover the code,
they use prerecorded input which may differ from your installation.
If that's the case (and the results are not what you expected), please
send me an example output of "readelf --debug-dump=line,info --wide"
(e.g. from Examples/StructureLayoutTest.cpp, see head of that source
for further details).

Feedback, additional debugger outputs or patches are very much
appreciated.

COPYRIGHT AND LICENCE

Copyright (C) 2007-2009 by Thomas Dorner (dorner (AT) cpan (DOT) org)

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.6.1 or,
at your option, any later version of Perl 5 you may have available.


Changes
Revision history for Perl extension Parse::Readelf.

0.17	(12.07.2013)
	fix for bad sort order when using display_nested_items

0.16	(06.04.2013-09.04.2013)
	Fix for multidimensional arrays

0.15	(26.12.2012-05.04.2013)
	added support for some more TAGs and attributes
	fixed missing include directive and added test in example source
	added first (experimental) support for Dwarf-4

0.14	(12.03.2012-03.10.2012)
	introduced constants for result tupel and fixed documentation (63778)

0.13	(12.03.2012)
	fixed bug with C typedefs (66776)

0.12	(09.03.2012)
	fixed tests for Carp >= 1.25

0.11	(26.11.2011)
	removed some intermediate paranoia checks

0.10	(15.03.2010-10.08.2011)
	fixed tests for MSWin without "cat"
	improved robustness of tail of expected error messages
	added possible wildcards and strings to recognition of const values

0.09	(21.01.2010)
	fixed bad results for empty regular expression in item_ids_matching

0.08	(21.12.2009)
	yet another missing "attribute to be ignored" found

0.07	(21.12.2009)
	fixed regexp for 1 byte locations
	added missing ignored attribute

0.06	(11.11.2009-20.12.2009)
	avoid SIGPIPE on close in tests 63++ of Parse-Readelf-Debug-Info.t
	fixed installation INC path in example structure_layout.pl
	fixed regexps to recognise changed output of newer readelf
	fixed sort order for structure layout
	added special handling of items with info for other items

0.05	(19.05.2009-30.05.2009)
	support of newer readelf/objdump versions
	example script now supports STDIN
	added support of volatile TAG, bugfix of postfixes

0.04	(05.04.2008-30.04.2009)
	hopefully fixed tests for Windows
	added Build.PL

0.03	(03.11.2007-10.12.2007)
	rebuild with newer MakeMaker, added missing ignored TAG

0.02	(27.09.2007-31.09.2007)
	fixed test to work with Perl version 5.6, added POD tests

0.01	(21.01.2007-10.09.2007)
	first version supporting printing of structure layout


lib/Parse/Readelf.pm
package Parse::Readelf;

# Author, Copyright and License: see end of file

=head1 NAME

Parse::Readelf - handle readelf's output with a class

=head1 SYNOPSIS

  use Parse::Readelf;

  my $readelf_data = new Parse::Readelf($executable);
  $readelf_data->print_structure_layout($re_identifier, 1);

=head1 ABSTRACT

Parse::Readelf parses (some of) the output of C<readelf> and stores
its interesting details in some objects to ease access.

At the moment only a very limited access to the structure layout of
data types and variables is supported.

=head1 DESCRIPTION

Normally an object of this class is constructed with the file name of
an object file to be parsed.  Upon construction the file is analysed
and all relevant information about its debug info section is stored
inside of the object or one of its subobjects.  This information can
be accessed afterwards using a bunch of getter methods, see
L</"METHODS"> for details.

This is BETA software, use at your own risk.

at the moment only information regarding the binary arrangement of
variables (Structure Layout) is supported (and that is regularly used
at my company, so the worst bugs should by found by now).  Other data
is ignored for now.

Currently only output for B<Dwarf version 2> is supported.  Please
contact the author for other versions and provide some example
C<readelf> outputs.

=cut

#########################################################################

use 5.006001;
use strict;
use warnings;
use Carp;

our $VERSION = '0.17';

use Parse::Readelf::Debug::Line;
use Parse::Readelf::Debug::Info ':constants';

#########################################################################

=head1 EXPORT

Nothing is exported by default as it's normally not needed to modify
the following variable:

This module exports nothing directly, it should be accessed via its
methods only.

=cut

# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #

require Exporter;

our @ISA = qw(Exporter);
our @EXPORT = qw();
our %EXPORT_TAGS = ( 'all' => [ qw(@structure_layout_types) ] );
our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

#########################################################################

=head2 I<@structure_layout_types>

is a list of the types that can be printed in a structure layout.  Its
elements are basically the tag identifieres from C<readelf>'s output
without the prefix B<DW_TAG_>.

=cut

# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #

our @structure_layout_types = qw(class
				 enumerat
				 member
				 structure
				 subrange
				 typedef
				 union
				 variable);

#########################################################################

=head2 new - get readelf's output into an object

    $readelf_data = new Parse::Readelf($file_name);

=head3 example:

    $readelf_data1 = new Parse::Readelf('program');
    $readelf_data2 = new Parse::Readelf('module.o');

=head3 parameters:

    $file_name          name of executable or object file

=head3 description:

    This method parses the output of several C<readelf> commands and
    stores its interesting details internally to be accessed later by
    getter methods described below.

=head3 returns:

    The method returns the blessed Parse::Readelf object or an
    exception in case of an error.

=cut

# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
sub new($$)
{
    my $this = shift;
    my $class = ref($this) || $this;
    my ($file_name) = @_;
    my %self = (line_info => undef,
		debug_info => undef);
    local $_;

    # checks:
    if (! $file_name)
    { croak 'bad call to new of ', __PACKAGE__ }
    if (ref($this))
    { carp 'cloning of a ', __PACKAGE__, ' object is not supported' }
    if (! -f $file_name)
    { croak __PACKAGE__, " can't find ", $file_name }

    # parse all supported readelf sections:
    $self{line_info} = new Parse::Readelf::Debug::Line($file_name);
    $self{debug_info} =
	new Parse::Readelf::Debug::Info($file_name, $self{line_info});

    # last consistency check:
    confess 'incomplete constructed object in ', __PACKAGE__
	unless defined $self{line_info} and defined $self{debug_info};

    bless \%self, $class;
}

#########################################################################

=head2 print_structure_layout - print structure layout of variables/types

    $readelf_data->print_structure_layout($re_name [, $print_location]);

=head3 example:

    $readelf_data->print_structure_layout('_t$');
    $readelf_data->print_structure_layout('_t$', 1);

=head3 parameters:

    $re_name            reg. exp. matching name of variable or data type
    $print_location     optional flag to print location with every definition

=head3 description:

    This method prints the structure layout of one or more variables
    or data types that match the regular expression for their name.
    If the optional parameter $print_location is true, each line also
    contains source location information, if availablble.

=head3 returns:

    nothing

=cut

# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
sub print_structure_layout($$;$)
{
    my $this = shift;
    my ($re_name, $print_location) = @_;
    local $_;

    # get item IDs for all matching items:
    my @ids =
	$this->{debug_info}
	    ->item_ids_matching($re_name,
				'^DW_TAG_(?:'.
				join('|', @structure_layout_types).
				')'
			       );

    # get layout for each item
    my @layouts = ();
    foreach (@ids)
    { push @layouts, $this->{debug_info}->structure_layout($_) }

    # get maximum width for each field:
    my ($level_name_width, $offset_width, $bit_offset_width, $type_width) =
	(1, 1, 0, 1);
    foreach (@layouts)
    {
	my $width = length($_->[$NAME]) + 2 * $_->[$LEVEL];
	$level_name_width = $width if $level_name_width < $width;
        $offset_width = length($_->[$OFFSET])
	    if $offset_width < length($_->[$OFFSET]);
        $bit_offset_width = length($_->[$BITOFFSET]) + 1
	    if  defined $_->[$BITOFFSET]  and
		$bit_offset_width < length($_->[$BITOFFSET]) + 1;
	$width = $_->[$TYPE] ne '' ? length($_->[$TYPE]) + 3 : 2;
	$width += length($_->[$SIZE]);
	$width += length(defined $_->[$BITSIZE]) + 4 if defined $_->[$BITSIZE];
	$type_width = $width if $width > $type_width;
    }

    # now print items:
    printf("%s   %-*s   %-*s   %s\n",
	   substr('OFFSET  ', 0 , $offset_width + $bit_offset_width),
	   $level_name_width, 'STRUCTURE',
	   $type_width, 'TYPE (SIZE)',
	   ($print_location ? 'SOURCE LOCATION' : ''));
    foreach (@layouts)
    {
	my $type_size = sprintf("%s(%s%d)",
				($_->[$TYPE] ne '' ? $_->[$TYPE].' ' : ''),
				(defined $_->[$BITSIZE]
				 ? $_->[$BITSIZE].' in ' : ''),
				$_->[$SIZE]);
	my $location = '';
	if ($print_location  and  defined $_->[$LOCATION])
	{
	    $location =
		$this->{line_info}->file($_->[$LOCATION]->[0],
					 $_->[$LOCATION]->[1],
					 1)
		    . ':' . $_->[$LOCATION]->[2];
	}
	printf("%0*d%-*s   %-*s   %-*s   %s\n",
	       $offset_width, $_->[$OFFSET],
	       $bit_offset_width, (defined $_->[$BITOFFSET]
				   ? '.'.$_->[$BITOFFSET] : ''),
	       $level_name_width, ('  ' x $_->[$LEVEL]) . $_->[$NAME],
	       $type_width, $type_size,
	       $location);
    }
}

1;

#########################################################################

__END__

=head1 KNOWN BUGS

Did I mentioned that this is Beta code?

Only Dwarf version 2 is supported.  Please contact the author for
other versions and provide some example C<readelf> outputs.

This has only be tested in a Unix like environment, namely Linux and
Solaris.

=head1 SEE ALSO

L<Parse::Readelf::Debug::Info>, L<Parse::Readelf::Debug::Line> and the
C<readelf> man page

=head1 AUTHOR

Thomas Dorner, E<lt>dorner (AT) cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2007-2013 by Thomas Dorner

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.6.1 or,
at your option, any later version of Perl 5 you may have available.

=cut


Examples/run_examples.pl
#!/usr/bin/perl
#
# Uses the script structure_layout.pl with different object files
# compiled from the example sources here.
#
# Author: Thomas Dorner
# Copyright: (C) 2007-2009 by Thomas Dorner (Artistic License)

use strict;
use warnings;

use File::Spec;

# make sure we are in the Examples directory
unless ($0 eq 'run_examples.pl')
{
    my @split_path = File::Spec->splitpath($0);
    my $path = File::Spec->catpath(@split_path[0..1]);
    chdir $path  or  die "$0: can't chdir to $path: $1\n";
}

# compile and parse example sources with different debugging formats:
my %source = (StructureLayoutTest => '.cpp');
my %output = (StructureLayoutTest =>
	      ['^sizeof\(Structure1\) == \d{1,2}$',
	       '^offsetof\(Structure1, m_04_pointer\) == \d{1,2}$',
	       '^sizeof\(l_pointer1\) == \d{1,2}$',
	       '^sizeof\(l_object1\) == \d{1,2}$',
	       '^sizeof\(l_object2a\) == \d{1,2}$',
	       '^sizeof\(l_object2b\) == \d{1,2}$',
	       '^sizeof\(l_cObject2b\) == \d{1,2}$',
	       '^sizeof\(l_object3\) == \d{1,2}$',
	       '^sizeof\(l_object4\) == \d{1,2}$',
	       '^sizeof\(l_object5\) == \d{1,2}$',
	       '^sizeof\(l_cvInt\) == \d{1,2}$',
	       '^sizeof\(l_objectU\) == \d{1,2}$',
	       '^sizeof\(l_object2_foo\) == \d{1,2}$',
	       '^sizeof\(l_object2_bar\) == \d{1,2}$']);
foreach my $base (keys %source)
{
 FORMAT:
    foreach my $format (qw(gdb stabs stabs+ coff xcoff xcoff+
			   dwarf-2 dwarf-4 vms))
    {
	foreach my $level (1..3)
	{
	    my $format_ok = 1;
	    my $object = $base.'-'.$format.'-'.$level.'.o';

	    # compile example source
	    my $command = 'g++ -Wall -Wextra -g'.$format.' -g'.$level;
	    $command .= ' -o'.$object.' -c '.$base.$source{$base};
	    open GCC, $command.' 2>&1 |'
		or  die  "$0: can't run '$command': $!\n";
	    while (<GCC>)
	    {
		if (m/ does not support .*$format.* format/)
		{
		    print 'Your system', $&, ".\n";
		    $format_ok = 0;
		}
	    }
	    close GCC  or  not $format_ok
		or  die  "$0: can't run '$command': $!";
	    next FORMAT unless $format_ok;

	    # link example source:
	    $command = 'g++ -o '.$base.' '.$object;
	    open GCC, $command.' 2>&1 |'
		or  die  "$0: can't run '$command': $!\n";
	    while (<GCC>)
	    {
		print $_;
		die 'a linker error should never happen';
	    }
	    close GCC  or  die  "$0: can't run '$command': $!";

	    # test example program:
	    $command = File::Spec->catfile('.', $base);
	    open PROG, $command.' 2>&1 |'
		or  die  "$0: can't run '$command': $!\n";
	    while (<PROG>)
	    {
 unless (defined $output{$base}->[$. - 1])
{
    die "$base,$output{$base},$.";
}
		my $match = $output{$base}->[$. - 1];
		unless (m/$match/)
		{
		    chomp;
		    print $base, ': "', $_, '" does not match /',
			$match, '/ (', $base, '-', $format, '-', $level, ")\n";
		    $format_ok = 0;
		}
	    }
	    close PROG  or  die  "$0: can't run '$command': $!";
	    next FORMAT unless $format_ok;

	    # bail out if not dwarf:
	    if ($format =~ m/^(?:stabs|x?coff|vms)/)
	    {
		print 'structure_layout.pl does not support "',
		    $format, "\" yet.\n";
		next FORMAT;
	    }

	    # test example program:
	    $command = 'perl -I'.File::Spec->catdir('..', 'lib');
	    $command .= ' '.File::Spec->catfile('structure_layout.pl');
	    $command .= ' "^S" '.$object.' >'.$object.'.layout';
	    system($command) == 0
		or  die  "$0: error running '$command': $?\n";
	    print 'Created example output "', $object, ".layout\".\n";
	}
    }
}


lib/Parse/Readelf/Debug/Line.pm
package Parse::Readelf::Debug::Line;

# Author, Copyright and License: see end of file

=head1 NAME

Parse::Readelf::Debug::Line - handle readelf's debug line section with a class

=head1 SYNOPSIS

  use Parse::Readelf::Debug::Line;

  my $line_info = new Parse::Readelf::Debug::Line($executable);

  my $object_id = $line_info->object_id("mocdule.c");

  my $file_name = $line_info->file($object_id, $number);
  my $directory_name = $line_info->directory($object_id, $number);
  my $path = $line_info->path($object_id, $number);

  my $object_name = $line_info->object_name($object_id);

  my $file_count = $line_info->files($object_id);
  my @files = $line_info->files($object_id);
  my $directory_count = $line_info->directories($object_id);
  my @directories = $line_info->directories($object_id);
  my $path_count = $line_info->paths($object_id);
  my @paths = $line_info->paths($object_id);

=head1 ABSTRACT

Parse::Readelf::Debug::Line parses the output of C<readelf
--debug-dump=line> and stores its interesting details in an object to
be available.  Normally it's not used directly but by other modules of
L<C<Parse::Readelf>>.

=head1 DESCRIPTION

Normally an object of this class is constructed with the file name of
an object file to be parsed.  Upon construction the file is analysed
and all relevant information about its debug line section is stored
inside of the object.  This information can be accessed afterwards
using a bunch of getter methods, see L</"METHODS"> for details.

Currently only output for B<Dwarf version 2> is supported.  Please
contact the author for other versions and provide some example
C<readelf> outputs.

=cut

#########################################################################

use 5.006001;
use strict;
use warnings;
use Carp;

our $VERSION = '0.17';

#########################################################################

=head1 EXPORT

Nothing is exported by default as it's normally not needed to modify
any of the variables declared in the following export groups:

=head2 :all

all of the following groups

=cut

# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #

require Exporter;

our @ISA = qw(Exporter);
our @EXPORT = qw();

our %EXPORT_TAGS =
    (command => [ qw($command) ],
     fixed_regexps => [ qw($re_section_start $re_dwarf_version) ],
     versioned_regexps => [ qw(@re_directory_table
			       @re_file_name_table
			       @re_file_name_table_header) ]
    );
$EXPORT_TAGS{all} = [ map { @$_ } values(%EXPORT_TAGS) ];

our @EXPORT_OK = ( @{ $EXPORT_TAGS{all} } );

#########################################################################

=head2 :command

=over

=item I<$command>

is the variable holding the command to run C<readelf> to get the
information relevant for this module, normally C<readelf
--debug-dump=line>.

=back

=cut

# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #

our $command = 'readelf --debug-dump=line';

#########################################################################

=head2 :fixed_regexps

=over

=item I<$re_section_start>

is the regular expression that recognises the start of the line debug
output of C<readelf>.

=item I<$re_dwarf_version>

is the regular expression that recognises the Dwarf version line in a
line debug output of C<readelf>.  The version number must be an
integer number which will (must) be stored in C<$1>.

=back

=cut

# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #

our $re_section_start =
    qr(^(?:raw )?dump of debug contents of section \.debug_line:)i;

our $re_dwarf_version = qr(^\s*DWARF Version:\s+(\d+)\s*$)i;

#########################################################################

=head2 :versioned_regexps

=over

=item I<@re_directory_table>

is the version dependent regular expression that recognises the start
of the directory table in line debug output of C<readelf>.

=item I<@re_file_name_table>

is the version dependent regular expression that recognises the start
of the non-empty file name table in line debug output of C<readelf>.

=item I<@re_file_name_table_header>

is the version dependent regular expression that recognises the
heading line of the file name table in line debug output of
C<readelf>.  If this must be modified this probably means the parsing
will not work correctly!

=back

=cut

# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #

our @re_directory_table =
    ( undef, undef, qr(^\s*The Directory Table)i );

our @re_file_name_table =
    (  undef, undef, qr(^\s*The File Name Table:)i );

our @re_file_name_table_header =
    (  undef, undef, qr(^\s*Entry\s+Dir\s+Time\s+Size\s+Name)i );

#########################################################################

=head2 new - get readelf's debug line section into an object

    $line_info = new Parse::Readelf::Debug::Line($file_name);

=head3 example:

    $line_info1 = new Parse::Readelf::Debug::Line('program');
    $line_info2 = new Parse::Readelf::Debug::Line('module.o');

=head3 parameters:

    $file_name          name of executable or object file

=head3 description:

    This method parses the output of C<readelf --debug-dump=line> and
    stores its interesting details internally to be accessed later by
    getter methods described below.

=head3 global variables used:

    The method uses all of the variables described above in the
    L</"EXPORT"> section.

=head3 returns:

    The method returns the blessed Parse::Readelf::Debug::Line object
    or an exception in case of an error.

=cut

# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
sub new($$)
{
    my $this = shift;
    my $class = ref($this) || $this;
    my ($file_name) = @_;
    my %self = (objects => [],
		object_map => {},
		directories => [],
		file_names => []);
    local $_;

    # checks:
    if (! $file_name)
    { croak 'bad call to new of ', __PACKAGE__; }
    if (ref($this))
    { carp 'cloning of a ', __PACKAGE__, " object is not supported"; }
    if (! -f $file_name)
    { croak __PACKAGE__, " can't find ", $file_name; }

    # call readelf and prepare parsing output:
    open READELF, '-|', $command.' '.$file_name  or
	croak "can't parse ", $file_name, ' with "', $command, '" in ',
	    __PACKAGE__, ': ', $!;

    # find start of section:
    while (<READELF>)
    { last if m/$re_section_start/; }

    # parse section:
    my $version = -1;
    my @directory_list = ();
    while (<READELF>)
    {

	if (m/$re_dwarf_version/)
	{
	    $version = $1;
	    confess 'DWARF version ', $version, ' not supported in ',
		__PACKAGE__
		    unless (defined $re_directory_table[$version]  and
			    defined $re_file_name_table[$version]  and
			    defined $re_file_name_table_header[$version]);
	}
	next unless $version >= 0;

	if (m/$re_directory_table[$version]/)
	{
	    @directory_list = ('.');
	    while (<READELF>)
	    {
		s/^\s+//; s/\s+$//;
		last unless $_;
		push @directory_list, $_;
	    }
	}

	elsif (m/$re_file_name_table[$version]/)
	{
	    <READELF> =~ m/$re_file_name_table_header[$version]/  or
		confess 'aborting: head line of file name table ',
		    'not recognised in ', __PACKAGE__;
	    my @file_name_table = ();
	    my @directory_table = ();
	    while (<READELF>)
	    {
		s/[\r\n]+//;
		last unless m/\s*(\d+)\s+(\d+)\s+(\d+)\s+(\d+)\s+(.*)/;
		my ($id, $directory_id, $time, $size, $name) =
		    ($1, $2, $3, $4, $5);
		if ($id == 1)
		{
		    push @{$self{objects}}, $name;
		    $self{object_map}{$name} = $#{$self{objects}};
		    push @{$self{directories}}, \@directory_table;
		    push @{$self{file_names}}, \@file_name_table;
		}
		$file_name_table[$id] = $name;
		$directory_table[$id] = $directory_list[$directory_id];
	    }
	    @directory_list = ();
	}
    }

    # now we're finished:
    close READELF  or
	croak 'error while attempting to parse ', $file_name,
	    ' (maybe not an object file?)';
    bless \%self, $class;
}

#########################################################################

=head2 object_id - get object ID of (named) source file

    $object_id = $line_info->object_id($file_name);

=head3 example:

    $object_id = $line_info->object_id('module.c');

=head3 parameters:

    $file_name          name of the source file (without directory)

=head3 description:

    This method returns the internal object ID of a module when given
    the name of its source file without directory.  This is a
    non-negative number.

=head3 returns:

    The method returns the object ID or -1 if no matching object was
    found.

=cut

# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
sub object_id($$)
{
    my $this = shift;
    my ($name) = @_;
    my $id = $this->{object_map}{$name};
    return defined $id ? $id : -1;
}

#########################################################################

=head2 object_name - get name of major source file for a given object ID

    $object_name = $line_info->object_name($object_id);

=head3 example:

    $object_name = $line_info->object_name(0);

=head3 parameters:

    $object_id          internal object ID of module

=head3 description:

    This method is the opposite method of L<|C<object_id>>, it returns
    the name of the major source file for the given internal object ID
    of a module.

=head3 returns:

    The method returns the source name or an empty string if no
    matching object was found.

=cut

# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
sub object_name($$)
{
    my $this = shift;
    my ($id) = @_;
    my $name = $this->{objects}[$id];
    return defined $name ? $name : '';
}

#########################################################################

=head2 file - get file name of source for a given ID combination

    $file_name = $line_info->file($object_id, $source_number, $relax);

=head3 example:

    $file_name = $line_info->file(0, 0);
    $file_name = $line_info->file(0, 0, 1); # Dwarf-4

=head3 parameters:

    $object_id          internal object ID of module
    $source_number      number of the source
    $relax              optional flag to enable fallback code for object ID

=head3 description:

    This method returns the file name (without directory) of the
    source file number C<$source_number> for the given internal object
    ID of a module.  The source number is a positive integer.  1 is
    the number of the major source file, all others are usually
    include files.  Note that 0 is not used!

    Newer Dwarf versions don't seem to use different tables for
    different object IDs and put all sources into one table.  The
    optional flag C<$relax> tells the method to use this one table in
    those cases.

=head3 returns:

    The method returns the source name or an empty string if no
    matching source was found in the object.

=cut

# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
sub file($$$;$)
{
    my $this = shift;
    my ($id, $source, $relax) = @_;
    # TODO: compilation unit and ID seem to be totally different
    # things and I've never seen 2 file name tables in Dwarf-4 so far:
    my $table = $this->{file_names}[$id];
    if (not defined $table  and  $relax)
    { $table = $this->{file_names}[0]; }
    return '' unless defined $table and ref($table) eq 'ARRAY';
    my $name = $table->[$source];
    return defined $name ? $name : '';
}

#########################################################################

=head2 files - list of all source file names for a given object ID

    @file_names = $line_info->files($object_id);
    $file_count = $line_info->files($object_id);

=head3 example:

    @file_names = $line_info->files(1);
    $number_of_files = $line_info->files($object_id);

=head3 parameters:

    $object_id          internal object ID of module

=head3 description:

    In list context this method returns a list of all file names
    (without directory parts) for the given internal object ID of a
    module.  In scalar context it returns how many elements this list
    would have.  As number 1 is the first source number actually used
    in the internal representation of the list the number returned in
    scalar context is also the last number you can pass to the
    L<|C<file>> method described above that returns a valid name (a
    non empty string).  Note also that the empty element 0 is not part
    of the list returned in list context.

=head3 returns:

    The method returns the list / the count as described above or an
    empty list / 0 if an unused or invalid object id was given.

=cut

# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
sub files($$)
{
    my $this = shift;
    my ($id) = @_;
    my $table = $this->{file_names}[$id];
    return wantarray ? () : 0 unless defined $table;
    if (wantarray)
    {
	return @{$table}[1..$#{$table}];
    }
    return $#{$table};
}

#########################################################################

=head2 directory - get directory name of source for a given ID combination

    $directory = $line_info->directory($object_id, $source_number);

=head3 example:

    $directory = $line_info->directory(0, 0);

=head3 parameters:

    $object_id          internal object ID of module
    $source_number      number of the source

=head3 description:

    This method returns the directory part of the file name of the
    source file number C<$source_number> for the given internal object
    ID of a module.  The source number is a positive integer.  1 is
    the number of the major source file, all others are usually
    include files.  Note that 0 is not used!

=head3 returns:

    The method returns the directory name or an empty string if no
    matching source was found in the object.

=cut

# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
sub directory($$$)
{
    my $this = shift;
    my ($id, $source) = @_;
    my $table = $this->{directories}[$id];
    return '' unless defined $table and ref($table) eq 'ARRAY';
    my $name = $table->[$source];
    return defined $name ? $name : '';
}

#########################################################################

=head2 directories - list of all directory names for a given object ID

    @directories = $line_info->directories($object_id);
    $dir_count = $line_info->directories($object_id);

=head3 example:

    @directories = $line_info->directories(1);
    $number_of_dirs = $line_info->directories($object_id);

=head3 parameters:

    $object_id          internal object ID of module

=head3 description:

    In list context this method returns a list of the directory parts
    of all file names for the given internal object ID of a module.
    As usually several used include files are found in the same
    directory this list normally will contain duplictes.  Those are NOT
    eliminated.  In scalar context it returns how many elements this
    list would have.  As number 1 is the first source number actually
    used in the internal representation of the list the number
    returned in scalar context is also the last number you can pass to
    the L<|C<directory>> method described above that returns a valid
    name (a non empty string).  Note also that the empty element 0 is
    not part of the list returned in list context.

=head3 returns:

    The method returns the list / the count as described above or an
    empty list / 0 if an unused or invalid object id was given.

=cut

# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
sub directories($$)
{
    my $this = shift;
    my ($id) = @_;
    my $table = $this->{directories}[$id];
    return wantarray ? () : 0 unless defined $table;
    if (wantarray)
    {
	return @{$table}[1..$#{$table}];
    }
    return $#{$table};
}

#########################################################################

=head2 path - get path to source file for a given ID combination

    $file_path = $line_info->path($object_id, $source_number);

=head3 example:

    $file_path = $line_info->path(0, 0);

=head3 parameters:

    $object_id          internal object ID of module
    $source_number      number of the source

=head3 description:

    This method returns the path (directory plus file name) of the
    source file number C<$source_number> for the given internal object
    ID of a module.  The source number is a positive integer.  1 is
    the number of the major source file, all others are usually
    include files.  Note that 0 is not used!

=head3 returns:

    The method returns the source name or an empty string if no
    matching source was found in the object.

=cut

# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
sub path($$$)
{
    my $this = shift;
    my ($id, $source) = @_;
    my $table = $this->{file_names}[$id];
    return '' unless defined $table  and  ref($table) eq 'ARRAY';
    my $name = $table->[$source];
    return '' unless defined $name;
    $table = $this->{directories}[$id];
    confess 'internal error: inconsistent table data for (',
	$id, ',', $source, ') in ', __PACKAGE__, '::path'
	    unless defined $table  and  ref($table) eq 'ARRAY';	# 1)
    my $name2 = $table->[$source];
    confess 'internal error: inconsistent name data for (',
	$id, ',', $source, ') in ', __PACKAGE__, '::path'
	    unless defined $name2; # 1)
    return $name2.'/'.$name;
}

#########################################################################

=head2 paths - list of paths to all sources for a given object ID

    @paths = $line_info->paths($object_id);
    $path_count = $line_info->paths($object_id);

=head3 example:

    @paths = $line_info->paths(1);
    $number_of_paths = $line_info->paths($object_id);

=head3 parameters:

    $object_id          internal object ID of module

=head3 description:

    In list context this method returns a list of all paths (directory
    plus file name) for the given internal object ID of a module.  In
    scalar context it returns how many elements this list would have.
    As number 1 is the first source number actually used in the
    internal representation of the list the number returned in scalar
    context is also the last number you can pass to the L<|C<file>>
    method described above that returns a valid name (a non empty
    string).  Note also that the empty element 0 is not part of the
    list returned in list context.

=head3 returns:

    The method returns the list / the count as described above or an
    empty list / 0 if an unused or invalid object id was given.

=cut

# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
sub paths($$)
{
    my $this = shift;
    my ($id) = @_;
    my $dir_table = $this->{directories}[$id];
    my $file_table = $this->{file_names}[$id];
    unless (defined $dir_table  and  defined $file_table)
    {
	confess 'internal error: inconsistent table data for (',
	    $id, ') in ', __PACKAGE__, '::paths'
		if defined $dir_table  or  defined $file_table;
	return wantarray ? () : 0;
    }
    confess 'internal error: inconsistent table structure for (',
	$id, ') in ', __PACKAGE__, '::paths'
	    unless (ref($dir_table) eq 'ARRAY'  and
		    ref($file_table) eq 'ARRAY');
    confess 'internal error: inconsistent name data for (',
	$id, ') in ', __PACKAGE__, '::paths'
	    unless $#{$dir_table} == $#{$file_table};
    return $#{$dir_table} unless wantarray();
    return
	map { $dir_table->[$_] . '/' . $file_table->[$_] }
	    (1..$#{$dir_table});
}

1;

#########################################################################

__END__

=head1 KNOWN BUGS

Only Dwarf version 2 is supported.  Please contact the author for
other versions and provide some example C<readelf> outputs.

This has only be tested in a Unix like environment and uses Unix path
syntax in some places.

=head1 SEE ALSO

L<Parse::Readelf> and the C<readelf> man page

=head1 AUTHOR

Thomas Dorner, E<lt>dorner (AT) cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2007-2013 by Thomas Dorner

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.6.1 or,
at your option, any later version of Perl 5 you may have available.

=cut
